ssl/quic/quic_tls.c

```diff
@@ -397,15 +389,30 @@ static int quic_read_record(OSSL_RECORD_LAYER *rl, void **rechandle,
     return OSSL_RECORD_RETURN_SUCCESS;
 }
 
-static int quic_release_record(OSSL_RECORD_LAYER *rl, void *rechandle)
+static int quic_release_record(OSSL_RECORD_LAYER *rl, void *rechandle,
+                               size_t length)
 {
-    if (!ossl_assert(rl->recread == 1) || !ossl_assert(rl == rechandle)) {
+    if (!ossl_assert(rl->recread > 0)
+            || !ossl_assert(rl->recunreleased <= rl->recread)
+            || !ossl_assert(rl == rechandle)
+            || !ossl_assert(length <= rl->recunreleased)) {
         QUIC_TLS_FATAL(rl, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
-        return 0;
+        return OSSL_RECORD_RETURN_FATAL;
+    }
+
+    rl->recunreleased -= length;
+
+    if (rl->recunreleased > 0)
+        return OSSL_RECORD_RETURN_SUCCESS;
+
+    if (!rl->qtls->args.crypto_release_rcd_cb(rl->recread,
```

Comment:
This would not be necessary if you called the ossl_rstream_release_record only with the size of the data that was processed and subsequently called ossl_rstream_get_record again to get pointer to the remaining data. Not sure if this helps in the way how the record layer is called from the TLS handshake layer or not.